s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala

```diff
@@ -648,47 +607,91 @@ import scala.collection.immutable
-        implicit val materializer = mat
-        import mat.executionContext
-        implicit val sys = mat.system
-        implicit val conf = resolveSettings(attr, mat.system)
+        implicit val conf: S3Settings = resolveSettings(attr, mat.system)
+        implicit val sys: ActorSystem = mat.system
+        implicit val materializer: Materializer = mat
+
+        val retriableFlow: Flow[(Chunk, (MultipartUpload, Int)), (Try[HttpResponse], (MultipartUpload, Int)), NotUsed] =
+          Flow[(Chunk, (MultipartUpload, Int))]
+            .map {
+              case (chunkedPayload, (uploadInfo, chunkIndex)) =>
+                //each of the payload requests are created
+                val partRequest =
+                  uploadPartRequest(uploadInfo, chunkIndex, chunkedPayload.data, chunkedPayload.size, headers)
+                (partRequest, (uploadInfo, chunkIndex))
+            }
+            .flatMapConcat { case (req, info) => Signer.signedRequest(req, signingKey).zip(Source.single(info)) }
+            .via(superPool[(MultipartUpload, Int)])
 
-        requestFlow
-          .via(superPool[(MultipartUpload, Int)])
-          .mapAsync(parallelism) {
-            case (Success(r), (upload, index)) =>
-              if (r.status.isFailure()) {
-                Unmarshal(r.entity).to[String].map { errorBody =>
-                  FailedUploadPart(
-                    upload,
-                    index,
-                    new RuntimeException(
-                      s"Upload part $index request failed. Response header: ($r), response body: ($errorBody)."
-                    )
-                  )
-                }
-              } else {
-                r.entity.discardBytes()
-                val etag = r.headers.find(_.lowercaseName() == "etag").map(_.value)
-                etag
-                  .map(t => Future.successful(SuccessfulUploadPart(upload, index, t)))
-                  .getOrElse(
-                    Future
-                      .successful(FailedUploadPart(upload, index, new RuntimeException(s"Cannot find etag in ${r}")))
-                  )
-              }
+        import conf.retrySettings._
 
-            case (Failure(e), (upload, index)) => Future.successful(FailedUploadPart(upload, index, e))
+        SplitAfterSize(chunkSize, chunkBufferSize)(atLeastOneByteString)
+          .via(getChunkBuffer(chunkSize, chunkBufferSize, maxRetriesPerChunk)) //creates the chunks
+          .concatSubstreams
+          .zip(requestInfo)
+          // Allow requests that fail with transient errors to be retried, using the already buffered chunk.
+          .via(RetryFlow.withBackoff(minBackoff, maxBackoff, randomFactor, maxRetriesPerChunk, retriableFlow) {
+            case (chunkAndUploadInfo, (Success(r), _)) =>
+              // 5xx errors from S3 can be treated as transient.
+              if (r.status.intValue >= 500) Some(chunkAndUploadInfo) else None
+            case (chunkAndUploadInfo, (Failure(_), _)) =>
```

Comment:
What failures do happen here? Are all of them worth retrying?

Reply:
So the only one I've encountered in recent memory is `akka.http.impl.engine.client.OutgoingConnectionBlueprint$UnexpectedConnectionClosureException`, which is basically a connection reset... something that S3 seems to do every so often in lieu of a 500.  My inclination is that any exception here is likely to be transient, and it's no great harm if it isn't (it'll still get thrown after a bit of a delay, and probably points to an environmental issue or something requiring human attention), particularly as UploadPart is idempotent.  I think if/when people do identify recurrent exceptions that can and do occur with some regularity, and where retrying is just a waste of time, they can add them here.